src/common/libterminus/client.c

```diff
@@ -0,0 +1,426 @@
+                              void *log_data)
+{
+    if (c) {
+        c->llog = log;
+        c->llog_data = log_data;
+    }
+}
+
+static void flux_pty_client_stop (struct flux_pty_client *c)
+{
+    flux_watcher_stop (c->fdw);
+    flux_watcher_stop (c->sw);
+}
+
+flux_future_t *flux_pty_client_wait (struct flux_pty_client *c)
+{
+    flux_future_t *f = flux_future_create (NULL, NULL);
+    if (flux_future_aux_set (f, "pty::client", c, NULL) < 0)
+        goto out;
+    if (c->h)
+        flux_future_set_flux (f, c->h);
+    if (zlist_append (c->waiters, f) < 0) {
+        flux_future_destroy (f);
+        f = NULL;
+    }
+out:
+    return f;
+}
+
+static void waiters_update_flux (struct flux_pty_client *c, flux_t *h)
+{
+    flux_future_t *f = zlist_first (c->waiters);
+    while (f) {
+        flux_future_set_flux (f, h);
+        f = zlist_next (c->waiters);
+    }
+}
+
+static int notify_exit (struct flux_pty_client *c, int status)
+{
+    flux_future_t *f = zlist_first (c->waiters);
+    while (f) {
+        flux_future_fulfill (f, NULL, NULL);
+        f = zlist_next (c->waiters);
+    }
+    return 0;
+}
+
+int flux_pty_client_get_wait_status (flux_future_t *f, int *statusp)
+{
+    struct flux_pty_client *c = flux_future_aux_get (f, "pty::client");
+    if (!c) {
+        errno = EINVAL;
+        return -1;
+    }
+    if (flux_future_get (f, NULL) < 0)
+        return -1;
+    if (statusp)
+        *statusp = c->wait_status;
+    return 0;
+}
+
+int flux_pty_client_exit_status (struct flux_pty_client *c)
+{
+    return c->wait_status;
+}
+
+static int flux_pty_client_set_server (struct flux_pty_client *c,
+                                       int rank,
+                                       const char *service)
+{
+    if (c == NULL) {
+        errno = EINVAL;
+        return -1;
+    }
+    if (!(c->service = strdup (service)))
+        return -1;
+    c->rank = rank;
+    return 0;
+}
+
+static void cls (void)
+{
+    /* ANSI clear screen + Home */
+    printf ("\033[2J\033[;H");
+    fflush (stdout);
+}
+
+void flux_pty_client_restore_terminal (void)
+{
+    if (term_needs_restoration) {
+        tcsetattr (STDIN_FILENO, TCSADRAIN, &orig_term);
+        /* https://en.wikipedia.org/wiki/ANSI_escape_code
+         * Best effort: attempt to ensure cursor is visible:
+         */
+        printf ("\033[?25h");
+        fflush (stdout);
+        term_needs_restoration = false;
+    }
+}
+
+static int setup_terminal (struct flux_pty_client *c)
+{
+    if (tcgetattr (STDIN_FILENO, &orig_term) < 0)
+        return -1;
+
+    c->term = orig_term;
+
+    cfmakeraw (&c->term);
+    c->term.c_cc[VLNEXT] = _POSIX_VDISABLE;
+    c->term.c_cc[VMIN] = 1;
+    c->term.c_cc[VTIME] = 0;
+    if (tcsetattr (STDIN_FILENO, TCSANOW, &c->term) < 0) {
+        llog_warning (c, "failed to setup terminal\n");
+        return -1;
+    }
+    term_needs_restoration = true;
+    atexit (flux_pty_client_restore_terminal);
+    cls ();
```

Comment:
any particular reason to clear the screen on pty?  I think it's more naturally to just fall into it on the current screen.  But perhaps that's just me.

Reply:
Hm, good question. My only answer is that this initially started as a clone of `dtach(1)`, which clears the screen when attaching to a new session. I'm guessing they mostly do this to give a visual cue that you've entered a new session, though I could imagine clearing the screen may help some cases when you put the terminal into raw mode. This also matches tmux/screen behavior so its what users are used to.

However, I just tried without the clear screen and it seems to work fine. Maybe somehow make this optional behavior?

I also learned that xterm has an alternate window buffer we could make use of here (So when you detach/exit tty contents would go back to what was on your screen before attach). I didn't go there now because it seemed superfluous.